### PR TITLE
Code cleanup sv_clipboard

### PR DESCRIPTION
**Pull request by dvdvideo1234**

### CODE REVIEW

File: lua/advdupe2/sv_clipboard.lua [hunk 1]
```diff
@@ -164,157 +212,153 @@ local function CopyEntTable( Ent, Offset )
...
-				Tab.BoneManip[ i ] = t
+		local t, s, a, p
+		local c = gtSetupTable.COMPARE
+		for i = 0, Ent:GetBoneCount() do
+			t = {}
+			s = Ent:GetManipulateBoneScale(i)
+			a = Ent:GetManipulateBoneAngles(i)
+			p = Ent:GetManipulateBonePosition(i)
+
+			-- Avoid making a vector just to compare it
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:251]: Use the setup compare lookup

File: lua/advdupe2/sv_clipboard.lua [hunk 2]
```diff
@@ -453,91 +494,95 @@ end
...
 
 		end
+
 		-- If there's a missing argument then unpack will stop sending at that argument
-		Val = Val or false
-		table.insert( Args, Val )
 
+		local Par = Val
+
+		if(Constraint.Type == "Slider" and Key == "material") then
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:579]: Patch the sliders and default the rope material to `cable/cable`

File: lua/advdupe2/sv_clipboard.lua [hunk 3]
```diff
@@ -640,56 +692,58 @@ local function CreateConstraintFromTable(Constraint, EntityList, EntityTable, Pl
...
-		if ( ModFunction ) then
-			status, error = pcall(ModFunction, Player, Ent, Data )
-			if(not status)then
-				if(Player)then
-					Player:ChatPrint('Error applying entity modifer, "'..tostring(Type)..'". ERROR: '..error)
+
+	for Type, Data in SortedPairs(Ent.EntityMods) do
+		local ModFunction = duplicator.EntityModifiers[Type]
+		if (ModFunction) then
+			local OK, ERR = pcall(ModFunction, Player, Ent, Data)
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:712]: Use local return values here

File: lua/advdupe2/sv_clipboard.lua [hunk 4]
```diff
@@ -640,56 +692,58 @@ local function CreateConstraintFromTable(Constraint, EntityList, EntityTable, Pl
...
 			end
 		end
 	end
-	if(Ent.EntityMods["mass"] and duplicator.EntityModifiers["mass"])then
-		status, error = pcall(duplicator.EntityModifiers["mass"], Player, Ent, Ent.EntityMods["mass"] )
-		if(not status)then
-			if(Player)then
-				Player:ChatPrint('Error applying entity modifer, "mass". ERROR: '..error)
+	if (Ent.EntityMods["mass"] and duplicator.EntityModifiers["mass"]) then
+		local OK, ERR = pcall(duplicator.EntityModifiers["mass"], Player, Ent, Ent.EntityMods["mass"])
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:723]: Here too

File: lua/advdupe2/sv_clipboard.lua [hunk 5]
```diff
@@ -1170,183 +1248,196 @@ local function AdvDupe2_Spawn()
...
-							if(v.Constraints~=nil)then
-								for i,c in pairs(v.Constraints)do
-									if(c and constraints[c.Type])then
-										edit=false
-										break
-									end
+			local phys
+			local edit
+			local mass
+			for k, v in pairs(Queue.CreatedEntities) do
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:1349]: Utilize the value of `k` instead of `_`

**Comment by dvdvideo1234** [comment]
@thegrb93 The file is ready for code review, thanks !

File: lua/advdupe2/sv_clipboard.lua [hunk 6]
```diff
@@ -546,74 +591,79 @@ local function CreateConstraintFromTable(Constraint, EntityList, EntityTable, Pl
...
+			if ConBuDupeInfo.Ent2Ang ~= nil then
+				second:SetAngles(ConBuDupeInfo.Ent2Ang)
+			elseif ConBuDupeInfo.Ent4Ang ~= nil then
+				second:SetAngles(ConBuDupeInfo.Ent4Ang)
+			end
 		end
 	end
 
-	local status, Ent = pcall( Factory.Func, unpack(Args))
+	local OK, Ent = pcall(Factory.Func, unpack(Args))
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:642]: Rename to `OK`

File: lua/advdupe2/sv_clipboard.lua [hunk 7]
```diff
@@ -829,148 +887,153 @@ end
...
+		for iNumber, Key in pairs(EntityClass.Args) do
 
 			Arg = nil
-			// Translate keys from old system
-			if ( Key == "pos" or Key == "position" ) then Key = "Pos" end
-			if ( Key == "ang" or Key == "Ang" or Key == "angle" ) then Key = "Angle" end
-			if ( Key == "model" ) then Key = "Model" end
-			if ( Key == "VehicleTable" and EntTable[Key] and EntTable[Key].KeyValues)then
-				EntTable[Key].KeyValues = {vehiclescript=EntTable[Key].KeyValues.vehiclescript, limitview=EntTable[Key].KeyValues.limitview}
+			-- Translate keys from old system
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:947]: Use the table lookup instead of comparing bunch of strings

File: lua/advdupe2/sv_clipboard.lua [hunk 8]
```diff
@@ -1047,106 +1111,119 @@ function AdvDupe2.duplicator.Paste( Player, EntityList, ConstraintList, Position
...
 			end
 
-			v:SetNotSolid( v.SolidMod )
+			v:SetNotSolid(v.SolidMod)
 		end
 	end
 	DisablePropCreateEffect = nil
-	hook.Call("AdvDupe_FinishPasting", nil, {{EntityList=EntityList, CreatedEntities=CreatedEntities, ConstraintList=ConstraintList, CreatedConstraints=CreatedConstraints, HitPos=OrigPos or Position, Player=Player}}, 1)
-	
+	hook.Call("AdvDupe_FinishPasting", nil, {
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:1164]: This looks a bit nicer

File: lua/advdupe2/sv_clipboard.lua [hunk 9]
```diff
@@ -1170,183 +1248,196 @@ local function AdvDupe2_Spawn()
...
 				end
-				undo.SetCustomUndoText("Undone "..(Queue.Name or "Advanced Duplication"))
-				undo.SetPlayer( Queue.Player )
+			end
+			undo.SetCustomUndoText("Undone " .. (Queue.Name or "Advanced Duplication"))
+			undo.SetPlayer(Queue.Player)
 			undo.Finish()
 
-			hook.Call("AdvDupe_FinishPasting", nil, {{EntityList=Queue.EntityList, CreatedEntities=Queue.CreatedEntities, ConstraintList=Queue.ConstraintList, CreatedConstraints=Queue.CreatedConstraints, HitPos=Queue.PositionOffset, Player=Queue.Player}}, 1)
+			hook.Call("AdvDupe_FinishPasting", nil, {
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:1419]: Make another hook pretty

**Comment by thegrb93** [comment]
This is almost too much to review. Please no more

**Comment by dvdvideo1234** [comment]
@thegrb93 Code style is not changed. The no-whitespace diff is [located here](https://github.com/wiremod/advdupe2/pull/316/files?w=1).

**Comment by thegrb93** [comment]
I know. It's still too much. I'll do when I get time.

File: lua/advdupe2/sv_clipboard.lua [hunk 10]
```diff
@@ -546,74 +591,79 @@ local function CreateConstraintFromTable(Constraint, EntityList, EntityTable, Pl
 	local Bone2
 	local Bone2Index
 	local ReEnableSecond
-	if(Constraint.BuildDupeInfo)then
+	local ConBuDupeInfo = Constraint.BuildDupeInfo
```
Comment (thegrb93) [inline lua/advdupe2/sv_clipboard.lua:594]: Rename this local `buildInfo`

File: lua/advdupe2/sv_clipboard.lua (hunk 6 above)
Comment (thegrb93) [reply lua/advdupe2/sv_clipboard.lua:642]: `ok` instead pls

File: lua/advdupe2/sv_clipboard.lua (hunk 3 above)
Comment (thegrb93) [inline lua/advdupe2/sv_clipboard.lua:712]: `err` instead of `ERR`

**Review by thegrb93 — CHANGES_REQUESTED** [review]

**Comment by thegrb93** [comment]
Mixing logic changes with hundreds of whitespace changes is extremely frowned upon. Don't do it again or I'll just close the PR.

The whitespace changes look fine except for the requested changes. The logic changes will need testing to make sure they work fine.

**Comment by dvdvideo1234** [comment]
Ok, I will turn off automatic lode alignment when do logic changes. Thanks!

File: lua/advdupe2/sv_clipboard.lua [hunk 11]
```diff
@@ -453,91 +494,95 @@ end
...
-									
-						end		
+
+						end
 					end
 
 				end
 
-				if Key == "Bone"..i or Key == "Bone" then Val = Constraint.Entity[ i ].Bone or 0 end
+				if Key == "Bone" .. i or Key == "Bone" then
```
[the remark this answers is not in the document]
Comment (dvdvideo1234) [reply lua/advdupe2/sv_clipboard.lua:543]: Doubt it.. What concerns me are those concatenations.. When you have `Bone13` or `Bonemanipulator1` they both start with `Bone`, but must handled differently. Same for `WPos%d` Maybe a pattern can be used to mach these. What's your `lookup_table.Pattern.bone = "Bone%d"` and later in the key `Key:find(lookup_table.Pattern.bone) or lookup_table.BONE[Key]`

File: lua/advdupe2/sv_clipboard.lua (hunk 3 above)
Comment (dvdvideo1234) [reply lua/advdupe2/sv_clipboard.lua]: Done in all places

File: lua/advdupe2/sv_clipboard.lua (hunk 10 above)
Comment (dvdvideo1234) [reply lua/advdupe2/sv_clipboard.lua]: Done.

File: lua/advdupe2/sv_clipboard.lua (hunk 11 above)
[the remark this answers is not in the document]
Comment (dvdvideo1234) [reply lua/advdupe2/sv_clipboard.lua:546]: I don't know however the concatenation or pattern matching is faster though.

**Comment by dvdvideo1234** [comment]
For `if(Constraint.Type == "Slider" and Key == "material") then` the `else` statement was used by default when you create a entity/constraint and a value is not present, the script fill it with false. So I decided to keep it and change only the sliders with a missing texture ( The value `false` is being defaulted by Garry to `cable/cable` and it and the slider creation function accepts string ). It is true, however when we make it throw error, it will show `the hidden rocks` when duping, so we can identify what is not being dupped with correct parameters in the first place. I think the original author went for the lazy approach putting `false` ;)

File: lua/advdupe2/sv_clipboard.lua [hunk 12]
```diff
@@ -591,57 +594,57 @@ local function CreateConstraintFromTable(Constraint, EntityList, EntityTable, Pl
...
+						Bone1:SetAngles(buildInfo.Bone1Angle)
 					end
 				end
 			end
 		end
 
 		if second ~= nil and not second:IsWorld() then
-			if ConBuDupeInfo.Ent2Ang ~= nil then
+			if buildInfo.Ent2Ang ~= nil then
 				second:SetAngles(ConBuDupeInfo.Ent2Ang)
```
Comment (thegrb93) [inline lua/advdupe2/sv_clipboard.lua:638]: Looks like you missed some `ConBuDupeInfo`

**Review by thegrb93 — CHANGES_REQUESTED** [review]

File: lua/advdupe2/sv_clipboard.lua (hunk 11 above)
[the remark this answers is not in the document]
Comment (thegrb93) [reply lua/advdupe2/sv_clipboard.lua:546]: Yeah nevermind. I'll refactor this function perhaps later.

File: lua/advdupe2/sv_clipboard.lua (hunk 2 above)
Comment (thegrb93) [reply lua/advdupe2/sv_clipboard.lua:582]: Can you delete this? It doesn't belong. https://github.com/wiremod/advdupe2/issues/319 will investigate replacing the 'false's to 'nil's.

**Review by thegrb93 — CHANGES_REQUESTED** [review]

File: lua/advdupe2/sv_clipboard.lua [hunk 13]
```diff
@@ -563,27 +579,22 @@ local function CreateConstraintFromTable(Constraint, EntityList, EntityTable, Pl
...
 		-- If there's a missing argument then unpack will stop sending at that argument
-
-		local Par = Val
-
-		if(Constraint.Type == "Slider" and Key == "material") then
-			Val = Val or "cable/cable"
-		else -- Force proper material data type string for sliders
-			Val = Val or false
-		end
+		Val = Val or false
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:597]: Reverted the old way the value is calculated

File: lua/advdupe2/sv_clipboard.lua [hunk 14]
```diff
@@ -105,6 +105,19 @@ local gtSetupTable = {
 		V1 = Vector(1, 1, 1),
 		A0 = Angle (0, 0, 0),
 		V0 = Vector(0, 0, 0)
+	},
```
Comment (dvdvideo1234) [inline lua/advdupe2/sv_clipboard.lua:108]: Added: Hash lookup to remove concatenations

File: lua/advdupe2/sv_clipboard.lua [hunk 15]
```diff
@@ -951,17 +962,19 @@ local function CreateEntityFromTable(EntTable, Player)
 			if (gtSetupTable.POS[Key]) then Key = "Pos" end
 			if (gtSetupTable.ANG[Key]) then Key = "Angle" end
 			if (gtSetupTable.MODEL[Key]) then Key = "Model" end
-			if (Key == "VehicleTable" and EntTable[Key] and EntTable[Key].KeyValues) then
```
Comment (thegrb93) [inline lua/advdupe2/sv_clipboard.lua]: Table lookups aren't faster than single string comparisons. Pls revert.

File: lua/advdupe2/sv_clipboard.lua [hunk 16]
```diff
@@ -105,6 +105,19 @@ local gtSetupTable = {
 		V1 = Vector(1, 1, 1),
 		A0 = Angle (0, 0, 0),
 		V0 = Vector(0, 0, 0)
+	},
+	TVEHICLE = {
+		["VehicleTable"] = true
+	},
+	SPECIAL = {
+		["VehicleTable"] = true
```
Comment (thegrb93) [inline lua/advdupe2/sv_clipboard.lua:113]: This key is wrong

File: lua/advdupe2/sv_clipboard.lua [hunk 17]
```diff
@@ -511,7 +527,7 @@ local function CreateConstraintFromTable(Constraint, EntityList, EntityTable, Pl
 
 		for i = 1, 4 do
 			if (Constraint.Entity and Constraint.Entity[i]) then
-				if Key == "Ent" .. i or Key == "Ent" then
+				if Key:find(gtSetupTable.FRM.E) then
```
Comment (thegrb93) [inline lua/advdupe2/sv_clipboard.lua:530]: Revert this please. This breaks the logic.

**Comment by thegrb93** [comment]
I'm getting tired of reviewing this. Stop adding changes.

**Comment by thegrb93** [comment]
Also your force-push means I have to look over the whole thing again.

File: lua/advdupe2/sv_clipboard.lua (hunk 17 above)
Comment (dvdvideo1234) [reply lua/advdupe2/sv_clipboard.lua]: Done.

File: lua/advdupe2/sv_clipboard.lua (hunk 16 above)
Comment (dvdvideo1234) [reply lua/advdupe2/sv_clipboard.lua]: Fixed.

**Comment by thegrb93** [comment]
Just needs testing now.

**Comment by dvdvideo1234** [comment]
@thegrb93 Cheers !

**Pull request merged**